### ISSUE TITLE
Java exceptions - When should we use it, and when not to?

### ISSUE THREAD

**Issue by zameschua** [nus-cs2103-AY1718S1/forum#34]
From what I read, exceptions should only handle 'unexpected' events, whereas IndexOutOfBounds should've been expected and checked.

From this, seems that many exception handlers in the 2K-LOC code should've been checked with if-else statements instead, ie.  IndexOutOfBounds exception in Command.getTargetPerson.

Or is my understanding of exceptions wrong?

**Comment by Zhiyuan-Amos** [comment]
You are right! ^^
Here's the documentation on Exceptions, hope it helps: https://docs.oracle.com/javase/tutorial/essential/exceptions/runtime.html

> Runtime exceptions represent problems that are the result of a programming problem, and as such, the API client code cannot reasonably be expected to recover from them or to handle them in any way.

> Here's the bottom line guideline: If a client can reasonably be expected to recover from an exception, make it a checked exception. If a client cannot do anything to recover from the exception, make it an unchecked exception.

As such, since we are expecting to recover from the exception, the API call should be throwing a checked exception. i.e. we can perhaps create our own Checked Exception `IndexOutOfBoundsCheckedException`, but this looks terribly out of the convention and it's a bad idea.

Which is why doing what you've suggested "check with if-else" statements is the right way to go!

**Comment by damithc** [comment]
I think @zameschua's question is regarding the code below (one example). He is asking whether it is better if we handle the 'invalid index' situation using normal if-else instead of using exceptions as done in the current code. @zameschua is that your question? In any case, I'll wait for other students to weigh in with their opinions.

```java
/**
 * Extracts the the target person in the last shown list from the given arguments.
 *
 * @throws IndexOutOfBoundsException if the target index is out of bounds of the last viewed listing
 */
protected ReadOnlyPerson getTargetPerson() throws IndexOutOfBoundsException {
    return relevantPersons.get(getTargetIndex() - DISPLAYED_INDEX_OFFSET);
}
```

**Comment by damithc** [comment]
Thanks for the inputs @Zhiyuan-Amos :+1: Good points. This is a bit subjective area. Let's see if there are differing opinions.

**Comment by Xenonym** [comment]
I feel that if it is with reference to the block of code that @damithc referred to, then the use of exceptions is justified, mainly because `getTargetPerson()` has no information on how to recover - only the caller would have information on why the index was out of bounds, and know how to recover.

More generally, my personal opinion (not at all a professional one) is that exceptions should be used when at the point of error, the function is unable to access information on what alternative actions it could take, and needs to defer to the first point at which action can be taken. Eg. in an `ArrayList`, the `get()` method throws an `IndexOutOfBounds` exception when an invalid index is given, because returning any object from the array would be incorrect behavior, and it has no idea why the index it was given was incorrect - only its caller would possibly know. Extending this, imagine that this invalid `get()` was called by a function to access items from storage based on a user given index. In this case, even though this code called `get()`, it *also* has no idea why this index was incorrect (after all, the index was user-supplied). Since returning any object would also be incorrect, the only course of action is to throw the exception up to the UI code, which can now handle the error by surfacing it to the user (and perhaps asking the user to correct their input).

**Comment by Zhiyuan-Amos** [comment]
> mainly because getTargetPerson() has no information on how to recover - only the caller would have information on why the index was out of bounds, and know how to recover.

@Xenonym what you've said here is right ^^ However, referring to the javadoc above, Unchecked / RunTime Exceptions such as `IndexOutOfBounds` Exception should not be handled. I.e. you shouldn't be doing something like:
```
try {
    list.get(index);
} catch (IndexOutOfBoundsException ioobe) {
    // handle exception
}
```

Rather, you should be doing something like:
```
if (index < 0 || index >= list.size()) {
    // handle "exception". You can throw an Exception here, not necessarily IndexOutOfBoundsException
}

object = list.get(index);
```

For e.g., you have a `DeleteCommand`, and you are calling `delete 10` where 10 is an out of bounds index. So in `DeleteCommand#execute(int)`, you can do something like:

```
public void execute(int index) throws ExecutionException {
    if (index < 0 || index >= list.size()) {
        throw new ExecutionException();
    }
    // delete person here
}
```

and `ExecutionException`: 
```
public class ExecutionException extends Exception {}
```

Take note that in the method header, declaration of `throws` should only be for checked Exceptions, not unchecked ones like `IndexOutOfBoundsException` i.e. the code in the code base that prof copied is not the right practice :P 

tl:dr what you said is correct; only callers know how to handle the exceptions thrown. However, we must differentiate between checked and unchecked Exceptions. Refer to the javadoc for more info! ^^

**Comment by Xenonym** [comment]
@Zhiyuan-Amos noted, thanks for the clarification re: checked and unchecked exceptions! However, I feel that the line between checked and unchecked exceptions seem a little unclear. If unchecked exceptions result from a programming problem, what makes something a programming problem, exactly? To take an example, IOException is a checked exception,  and can be thrown if you try to open a non-existent file. On the one hand, we can argue that it is not the fault of the programmer that the file doesn't exist. On the other hand, we could also argue that it is the programmer's fault for not checking that the file exists before trying to open it, and hence should be an unchecked exception. IndexOutOfBounds similarly could also be argued to be recoverable (since we can always try another index), and hence should be checked. Is there something regarding the difference that I am missing?

**Comment by Zhiyuan-Amos** [comment]
<strike>
From javadoc: 
> One case where it is common practice to throw a RuntimeException is when the user calls a method incorrectly. For example, a method can check if one of its arguments is incorrectly null. If an argument is null, the method might throw a NullPointerException, which is an unchecked exception.

When is a method called incorrectly? Hint: Think preconditions. :)
</strike>

Just ignore this and read below!

**Comment by Zhiyuan-Amos** [comment]
I thought it may be confusing so I decided to reply instead ^^

**Main point**

For the example you've given about `IOException`, from this link: https://stackoverflow.com/questions/13216148/java-what-throws-an-ioexception, it seems that the more common examples of an `IOException` aren't programmer errors. As such, that perhaps is the reason why `IOException` is a checked Exception. One of the examples given "Reading a network file and got disconnected." -> It's not a programmer's fault that this happened.

On another hand, concerning `IndexOutOfBoundsException`, honestly, when you took (or if you took) CS1010 and CS1020, when you did your loops and your program crashed cos of `IndexOutOfBoundsException` or `NullPointerException`, it's your fault cos you looped out of bounds of the array or you didn't create a new object :P
______________

**Additional notes: LONG READ**

Also, what you mentioned above is perfectly right: We can argue that we want to recover from an `IndexOutOfBoundsException` some times. So how? :O Should we make it an unchecked or checked Exception? This is kinda gray isn't it. **In Software Engineering, there's this idea of trade-off: we weigh the pros and cons and then decide.**

If `IndexOutOfBoundsException` (or even `NullPointerException`, because we can argue that we want to recover from a null value too) is a checked Exception, then your code will look horribly long because you will have things like:

```
try {
    list.set(index, object);
} catch (NullPointerException npe) {
    // do something
} catch (IndexOutOfBoundsException ioobe) {
    // do something else
} catch (ClassCastException cce) {
    // do something else
}
```

Or let's say if you want to handle them all the same, you still have to do:
```
try {
    list.set(index, object);
} catch (Exception e) {
    // do something
}
```

And many Java API call has Exceptions involved. If everything were to be checked Exceptions, your code will look horribly long. (https://docs.oracle.com/javase/8/docs/api/java/util/List.html#set-int-E- List#set(int, Object) has 5 Exceptions LOL. The example given above isn't made up on my own).

There are definitely some times when you want to handle some unchecked Exceptions like `IndexOutOfBoundsException`, but I guess these times aren't that often, and **for the sake of making code look less verbose, it's a necessary trade-off.** From the java docs: "Runtime exceptions can occur anywhere in a program, and in a typical one they can be very numerous. Having to add runtime exceptions in every method declaration would reduce a program's clarity. Thus, the compiler does not require that you catch or specify runtime exceptions (although you can)." The context is slightly different though; in our context, **making unchecked exceptions to become checked exceptions will reduce a program's clarity, making it a bad trade-off.**

Edit:
Also, again back to my main point, since these Exceptions are **likely** to be caused by programmer errors, that's why these are unchecked Exceptions. I know in AddressBook, things like `DeleteCommand` can handle `IndexOutOfBoundsException` by asking for user input. However, there are many other times when you crash into `IndexOutOfBoundsException` not because of user input, but because of your own programming errors (like CS1010 & CS1020 :P). So, it's a good trade-off to make it an unchecked Exception, rather than making it a checked Exception and having everyone to handle it.

**Comment by Xenonym** [comment]
Wow thanks so much for the detailed reply @Zhiyuan-Amos! Seems like there is only so much we can systematize - there's always gonna be that element of human judgement that comes into play.

**Comment by Zhiyuan-Amos** [comment]
> Seems like there is only so much we can systematize - there's always gonna be that element of human judgement that comes into play.

@Xenonym yes you are right! ^^ There'll be some gray areas for sure. Don't worry too much about it; this kind of qn won't come out for exams (at least it didn't come out for mine). Am I right @damithc :stuck_out_tongue_closed_eyes:?
